File: src/Problem017NumberLetterCounts.java
private static final Map<Integer, String> numberNames = new HashMap<>() {{
  put(0, "");
  put(1, "one");
  put(2, "two");
  put(3, "three");
  put(4, "four");
  put(5, "five");
  put(6, "six");
  put(7, "seven");
  put(8, "eight");
  put(9, "nine");
  put(10, "ten");
  put(11, "eleven");
  put(12, "twelve");
  put(13, "thirteen");
  put(14, "fourteen");
  put(15, "fifteen");
  put(16, "sixteen");
  put(17, "seventeen");
  put(18, "eighteen");
  put(19, "nineteen");
  put(20, "twenty");
  put(30, "thirty");
  put(40, "forty");
  put(50, "fifty");
  put(60, "sixty");
  put(70, "seventy");
  put(80, "eighty");
  put(90, "ninety");
  put(100, "hundred");
  put(1000, "one thousand");
}};

public static String writeNumber(int n) {
  if (n == 0) return "";
  if (n <= 20 || n == 1000) return numberNames.get(n);
  if (n < 100) {
    int decimalDigit = n / 10;
    int decimalSentence = decimalDigit * 10;
    int unitDigit = n - decimalSentence;
    return numberNames.get(decimalSentence) + (unitDigit > 0 ? "-" : "") + numberNames.get(unitDigit);
  }
  if (n < 1000) {
    int hundredDigit = n / 100;
    int hundredSentence = hundredDigit * 100;
    int decimalPart = n - hundredSentence;
    String hundredText = numberNames.get(hundredDigit) + " hundred";
    return hundredText + (decimalPart > 0 ? " and " : "") + writeNumber(decimalPart);
  }
  throw new IllegalArgumentException("Invalid number");
}

public static int countLetters(String input) {
  String onlyLetters = input.replaceAll("[^A-Za-z]", "");
  return onlyLetters.length();
}

void main() {
  int totalChars = 0;

  for (int i = 1; i <= 1000; i++) {
    String writtenNumber = writeNumber(i);
    int chars = countLetters(writtenNumber);
    totalChars += chars;
    //IO.println("[" + i + "] -> " + writtenNumber + " (" + chars + ")");
  }
  IO.println("Total letters were written: " + totalChars);
}

File: src/Problem017NumberLetterCounts.kt
val numberNames:Map<Int, String> = mutableMapOf(
  0 to "",
  1 to "one",
  2 to "two",
  3 to "three",
  4 to "four",
  5 to "five",
  6 to "six",
  7 to "seven",
  8 to "eight",
  9 to "nine",
  10 to "ten",
  11 to "eleven",
  12 to "twelve",
  13 to "thirteen",
  14 to "fourteen",
  15 to "fifteen",
  16 to "sixteen",
  17 to "seventeen",
  18 to "eighteen",
  19 to "nineteen",
  20 to "twenty",
  30 to "thirty",
  40 to "forty",
  50 to "fifty",
  60 to "sixty",
  70 to "seventy",
  80 to "eighty",
  90 to "ninety",
  100 to "hundred",
  1000 to "one thousand"
)

fun writeNumber(n: Int): String {
  if (n == 0) return ""
  if (n <= 20 || n == 1000) {
    return numberNames[n]!!
  }

  if (n < 100) {
    val decimalDigit = n / 10;
    val decimalSentence = decimalDigit * 10;
    val unitDigit = n - decimalSentence;

    return "${numberNames[decimalSentence]}${ if (unitDigit > 0) "-" else ""}${numberNames[unitDigit]}"
  }

  if (n < 1000) {
    val hundredDigit = n / 100
    val hundredSentence = hundredDigit * 100
    val decimalPart = n - hundredSentence

    val hundredText = "${numberNames[hundredDigit]} hundred"

    return "$hundredText${if(decimalPart > 0) " and " else ""}${writeNumber(decimalPart)}"
  }

  throw Error("Invalid number")
}

fun countLetters(input: String): Int {
  return Regex("[a-zA-Z]").findAll(input).count()
}

fun main() {
  var totalChars = 0
  (1 until 1001).asSequence()
    .forEach {
      val writtenNumber = writeNumber(it)
      val chars = countLetters(writtenNumber)
      totalChars += chars
      println("[$it] -> ${writtenNumber} (${chars})")
    }
  println("Total letters were written: $totalChars")
}
